### PR TITLE
Add Mixed16 tests to multi-node chain List

### PR DESCRIPTION
**Pull request by belldandyxtq**
This commit adds mixed16 tests to multi-node chain list

Thank you for creating a pull request!

Please double-check the following.

- Read [our contribution guide](https://docs.chainer.org/en/stable/contribution.html).
  - Does your code conform to our coding guidelines?
  - Did you write sufficient test code?
  - Did you write sufficient documentation?
- Also, take a look at [our compatibility policy](https://docs.chainer.org/en/stable/compatibility.html).

### CODE REVIEW

**Comment by belldandyxtq** [comment]
pfnci, test this please

**Comment by belldandyxtq** [comment]
pfnci, test this please

File: tests/chainermn_tests/links_tests/test_multi_node_chain_list.py [hunk 1]
```diff
@@ -238,8 +240,9 @@ def check_cycle_model(gpu, param):
     n, d = 100, 10
 
     with chainer.using_config('dtype', param.dtype):
+        io_dtype = chainer.get_dtype(map_mixed16=np.float16)
```
Comment (kuenishi) [inline tests/chainermn_tests/links_tests/test_multi_node_chain_list.py]: What is the intention here on explicitly setting `map_mixed16` ? I'm afraid we don't need this line.

File: tests/chainermn_tests/links_tests/test_multi_node_chain_list.py [hunk 2]
```diff
@@ -279,11 +282,13 @@ def test_cycle_model_gpu(param):
 def check_crossing_model(gpu, param):
     communicator, rank_next, rank_prev = create_communicator(gpu)
 
-    n, d = 100, 10
-    X = np.random.randn(n, d).astype(param.dtype)
-    Y = (np.random.rand(n) * 2).astype(np.int32)
-
     with chainer.using_config('dtype', param.dtype):
+        n, d = 100, 10
+        io_dtype = chainer.get_dtype(map_mixed16=np.float16)
```
Comment (kuenishi) [inline tests/chainermn_tests/links_tests/test_multi_node_chain_list.py]: ditto

File: tests/chainermn_tests/links_tests/test_multi_node_chain_list.py [hunk 3]
```diff
@@ -317,10 +322,13 @@ def test_crossing_model_gpu(param):
 def check_branching_model(gpu, communicator, rank_next, rank_prev,
                           parent_model, param):
     n, d = 100, 10
-    X = np.random.randn(n, d).astype(param.dtype)
-    Y = (np.random.rand(n) * 2).astype(np.int32)
 
     with chainer.using_config('dtype', param.dtype):
+        io_dtype = chainer.get_dtype(map_mixed16=np.float16)
```
Comment (kuenishi) [inline tests/chainermn_tests/links_tests/test_multi_node_chain_list.py]: ditto

File: tests/chainermn_tests/links_tests/test_multi_node_chain_list.py [hunk 4]
```diff
@@ -279,11 +282,13 @@ def test_cycle_model_gpu(param):
...
-    n, d = 100, 10
-    X = np.random.randn(n, d).astype(param.dtype)
-    Y = (np.random.rand(n) * 2).astype(np.int32)
-
     with chainer.using_config('dtype', param.dtype):
+        n, d = 100, 10
+        io_dtype = chainer.get_dtype(map_mixed16=np.float16)
+
+        X = np.random.randn(n, d).astype(io_dtype)
+        Y = (np.random.rand(n) * 2).astype(np.int32)
```
Comment (kuenishi) [inline tests/chainermn_tests/links_tests/test_multi_node_chain_list.py]: Why are these lines moved here?

File: tests/chainermn_tests/links_tests/test_multi_node_chain_list.py (hunk 1 above)
Comment (belldandyxtq) [reply tests/chainermn_tests/links_tests/test_multi_node_chain_list.py]: Because the mixed16 type is a chainer type, which doesn't exist in numpy. To create a numpy object, mixed16 needs to be maped to a numpy type, which is this line

File: tests/chainermn_tests/links_tests/test_multi_node_chain_list.py (hunk 4 above)
Comment (belldandyxtq) [reply tests/chainermn_tests/links_tests/test_multi_node_chain_list.py]: The chainer.get_type needs to be used under a context

File: tests/chainermn_tests/links_tests/test_multi_node_chain_list.py (hunk 1 above)
Comment (belldandyxtq) [reply tests/chainermn_tests/links_tests/test_multi_node_chain_list.py]: It is fine to manually identify the mixed16 and set to nump.float16. Since chainer has already provided this functionality, it is more chaineric to use chainer.get_dtype

Comment (kuenishi) [reply tests/chainermn_tests/links_tests/test_multi_node_chain_list.py]: mixed16 doesn't need to be mapped if we intend fp16.

```python
>>> import numpy, chainer
>>> numpy.arange(10).reshape(1, 10).astype(chainer.mixed16)
array([[0., 1., 2., 3., 4., 5., 6., 7., 8., 9.]], dtype=float16)
```

**Comment by kuenishi** [comment]
Jenkins, test this please.
